File: sangsoo/0xFF ETC/oop.cpp
#include<iostream>
using namespace std;

// Encapsulation
class Encapsulation {
private:
    int x;

public:
    void set(int a) {
        x = a;
    }

    int get() {
        return x;
    }
};

// Inheritance
class Parent {
public:
    void parentFunction() {
        cout << "This is a function in the Parent class." << endl;
    }
};

class Child : public Parent {
public:
    void childFunction() {
        cout << "This is a function in the Child class." << endl;
    }
};

// Polymorphism
class Base {
public:
    virtual void show() { 
        cout << "Base class\n"; 
    }
};

class Derived : public Base {
public:
    void show() { 
        cout << "Derived class\n"; 
    }
};

// Abstract class for Abstraction
class AbstractClass {
public:
    virtual void pureVirtualFunction() = 0;
};

class ConcreteClass : public AbstractClass {
public:
    void pureVirtualFunction() {
        cout << "Implemented Virtual Function" << endl;
    }
};

int main() {
    // Encapsulation
    Encapsulation obj;
    obj.set(5);
    cout << "Encapsulation: " << obj.get() << "\n";

    // Inheritance
    Child child;
    child.parentFunction();
    child.childFunction();

    // Polymorphism
    Base* base; 
    Derived derived;
    base = &derived;  
    base->show();

    // Abstraction
    ConcreteClass concrete;
    concrete.pureVirtualFunction();

    return 0;
}

File: sangsoo/0xFF ETC/oop.py
# Encapsulation
class Encapsulation:
    def __init__(self):
        self.__private_var = None  

    def set_var(self, value):
        self.__private_var = value

    def get_var(self):
        return self.__private_var

encapsulation = Encapsulation()
encapsulation.set_var(5)
print(f'Encapsulation: {encapsulation.get_var()}')

# Inheritance
class Parent:
    def parent_function(self):
        print('This is a function in the Parent class.')

class Child(Parent):
    def child_function(self):
        print('This is a function in the Child class.')

child = Child()
child.parent_function()
child.child_function()

# Polymorphism
class Base:
    def show(self):
        print('Base class')

class Derived(Base):
    def show(self):
        print('Derived class')

derived = Derived()
derived.show()

# Abstraction
from abc import ABC, abstractmethod

class AbstractClass(ABC):
    @abstractmethod
    def abstract_method(self):
        pass

class ConcreteClass(AbstractClass):
    def abstract_method(self):
        print('Implemented abstract method')

concrete = ConcreteClass()
concrete.abstract_method()
